sandbox/cgroup/tracking.go

```diff
@@ -0,0 +1,243 @@
+	pid := osGetpid()
+	path, err := cgroupProcessPathInTrackingCgroup(pid)
+	if err != nil {
+		return err
+	}
+	unitName := fmt.Sprintf("%s.service", securityTag)
+	if !strings.Contains(path, unitName) {
+		return ErrCannotTrackProcess
+	}
+	return nil
+}
+
+func sessionOrMaybeSystemBus(uid int) (isSessionBus bool, conn *dbus.Conn, err error) {
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus. We have a preference for session bus, as this is
+	// where applications normally go to. When a session bus is not available
+	// and the invoking user is root, we use the system bus instead.
+	//
+	// It is worth noting that hooks will not normally have a session bus to
+	// connect to, as they are invoked as descendants of snapd, and snapd is a
+	// service running outside of any session.
+	conn, err = dbusutil.SessionBus()
+	if err == nil {
+		logger.Debugf("using session bus")
+		return true, conn, nil
+	}
+	logger.Debugf("session bus is not available: %s", err)
+	if uid == 0 {
+		logger.Debugf("falling back to system bus")
+		conn, err = dbusutil.SystemBus()
+		if err != nil {
+			logger.Debugf("system bus is not available: %s", err)
+		} else {
+			logger.Debugf("using system bus now, session bus was not available")
+		}
+	}
+	return isSessionBus, conn, err
+}
+
+var errDBusUnknownMethod = errors.New("org.freedesktop.DBus.Error.UnknownMethod")
+var errDBusNameHasNoOwner = errors.New("org.freedesktop.DBus.Error.NameHasNoOwner")
+var errDBusSpawnChildExited = errors.New("org.freedesktop.DBus.Error.Spawn.ChildExited")
+
+// doCreateTransientScope creates a systemd transient scope with specified properties.
+//
+// The scope is created by asking systemd via the specified DBus connection.
+// The unit name and the PID to attach are provided as well. The DBus method
+// call is performed outside confinement established by snap-confine.
+var doCreateTransientScope = func(conn *dbus.Conn, unitName string, pid int) error {
+	// The property and auxUnit types are not well documented but can be traced
+	// from systemd source code. Systemd defines the signature of
+	// StartTransientUnit as "ssa(sv)a(sa(sv))". The signature can be
+	// decomposed as follows:
+	//
+	// Partial documentation, at the time of this writing, is available at
+	// https://www.freedesktop.org/wiki/Software/systemd/dbus/
+	//
+	// unitName string // name of the unit to start
+	// jobMode string  // corresponds to --job-mode= (see systemctl(1) manual page)
+	// properties []struct{
+	//   Name string
+	//   Value interface{}
+	// } // properties describe properties of the started unit
+	// auxUnits []struct {
+	//   Name string
+	//   Properties []struct{
+	//   	Name string
+	//   	Value interface{}
+	//	 }
+	// } // auxUnits describe any additional units to define.
+	type property struct {
+		Name  string
+		Value interface{}
+	}
+	type auxUnit struct {
+		Name  string
+		Props []property
+	}
+
+	mode := "fail"
```

Comment:
can you add a comment about what this mode exactly mean?

Reply:
Indeed, this is a bit mysterious part of the API. Will do!